### PR TITLE
Fix #2836: Disallow @JSName as loading spec.

### LINKED ISSUE #2836: Disallow @JSName in favor of @JSGlobal

### PR DESCRIPTION
**Pull request by gzm0**
As a side effect, this enforces having a loading spec for JS native classes.

### CODE REVIEW

**Comment by gzm0** [comment]
Review by @sjrd

File: compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala [hunk 1]
```diff
@@ -574,10 +574,12 @@ abstract class PrepJSInterop extends plugins.PluginComponent
           for {
             annot <- sym.annotations
             annotSym = annot.symbol
-            if JSNativeLoadingSpecAnnots.contains(annotSym)
```
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala]: If you move the guard to the body, consider also moving `annotSym` as a `val annotSym = ...` in the body of the for. It's a bit weird to have the last line of a for be an `=`.

File: compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala [hunk 2]
```diff
@@ -675,13 +677,13 @@ abstract class PrepJSInterop extends plugins.PluginComponent
...
-          if JSNativeLoadingSpecAnnots.contains(annotSym)
         } {
-          if (annotSym != JSNameAnnotation) {
+          if (JSNativeLoadingSpecAnnots.contains(annotSym)) {
             reporter.error(annot.pos,
                 "Classes and objects nested in a JS native object cannot " +
                 s"have an @${annotSym.nameString} annotation.")
-          } else if (annot.args.head.tpe.typeSymbol != StringClass) {
+          } else if (annotSym == JSNameAnnotation &&
+                annot.args.head.tpe.typeSymbol != StringClass) {
```
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala]: Indentation issue. This should be indented 4 spaces wrt. `}`

File: compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala [hunk 3]
```diff
@@ -746,45 +737,11 @@ abstract class PrepJSInterop extends plugins.PluginComponent
...
-                    "Top-level native JS classes and objects should have an " +
-                    "@JSGlobal or @JSImport annotation. This will be " +
-                    "enforced in 1.0.\n" +
-                    "  If migrating from 0.6.14 or earlier, the equivalent " +
-                    "behavior is an @JSGlobal without parameter." +
-                    SuppressMissingJSGlobalDeprecationsMsg)
-              }
-            }
+            reporter.error(pos, "Native JS classes and objects must have an " +
+                "@JSGlobal or @JSImport annotation")
```
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala]: Or `@JSGlobalScope`. Consider a slightly different wording:

    Native JS classes and objects must have exactly one annotation among @JSGlobal, @JSImport and @JSGlobalScope

It reads a bit better with 3 alternatives.

Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala]: Also consider moving this check inside `checkAndGetJSNativeLoadingSpecAnnotOf`, and unify the error message with the case with duplicate loading spec annots.

File: compiler/src/test/scala/org/scalajs/core/compiler/test/JSInteropTest.scala [hunk 4]
```diff
@@ -303,25 +345,12 @@ class JSInteropTest extends DirectTest with TestHelpers {
     } {
       val expectedMessage = {
         s"""
-          |newSource1.scala:7: error: Native JS classes and objects can only have one annotation among JSName, JSGlobal, JSImport and JSGlobalScope.
+          |newSource1.scala:7: error: Native JS classes and objects can only have one annotation among JSGlobal, JSImport and JSGlobalScope.
           |$secondAnnot
           | ^
         """
```
Comment (sjrd) [inline compiler/src/test/scala/org/scalajs/core/compiler/test/JSInteropTest.scala]: Consider inlining the rhs of `expectedMessage` where it is used, as an argument to `hasErrors`, now that it doesn't need to be composed with `jsNameWarning` anymore.

**Comment by gzm0** [comment]
Updated and rebased.

**Comment by sjrd** [comment]
Jenkins, retest this please

**Pull request merged**